### PR TITLE
Speed up SGPR covariances

### PR DESCRIPTION
**Pull request by gpleiss**
Addresses problem in #1515 . @monabf - can you check out the `faster_sgpr_covar` branch and make sure that this fix works for you? If so, then we'll cut a bug fix release that should hopefully fix these SGPR issues once and for all :)

### CODE REVIEW

**Comment by monabf** [comment]
Thanks a lot for looking into this @gpleiss! I'm checking it out today, I'll get back to you as soon as I can :)

**Comment by monabf** [comment]
Hey @gpleiss, so I checked this out. It seems like the last problem I was mentioning (the one that reopened the issue) is indeed solved, predictions seem consistently fast. Thanks a lot for trying to fix once and for all the SGPR issues I've been facing :)

However, I still have a problem with the predictive variance. The following script fits a sparse GP model to toy 1d data. It found a good fit and a good predictive variance 10/10 times before the GPyTorch 1.4.0 release, 9/10 times after your fix of #1515, and now 0/10 times. I'm not really sure where this comes from since my usual test script with the elevator data (the one I've been posting in all my issues) seems ok. Hope you have an idea!

```python
import torch
import gpytorch
import math, time

from gpytorch.kernels import InducingPointKernel
from matplotlib import pyplot as plt

# Set double precision by default
torch.set_default_tensor_type(torch.DoubleTensor)
torch.set_default_dtype(torch.float64)

class SparseGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, kernel,
                 inducing_points):
        train_x = torch.squeeze(train_x)
        train_y = torch.squeeze(train_y)
        train_y = train_y.contiguous()
        super(SparseGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.base_covar_module = kernel
        self.covar_module = InducingPointKernel(
            self.base_covar_module, inducing_points=inducing_points,
            likelihood=likelihood)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

    def optimize(self, likelihood, train_x, train_y, training_iter=50,
                 optimizer=torch.optim.Adam, lr=0.1, parameters=None,
                 verbose=False):
        train_x = torch.squeeze(train_x).double()
        train_y = torch.squeeze(train_y).double()
        # https://docs.gpytorch.ai/en/v1.2.0/examples/00_Basic_Usage/Hyperparameters.html
        self.train()
        likelihood.train()
        if not parameters:
            parameters = self.parameters()
        optimizer = optimizer(parameters, lr=lr)
        mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, self)
        for i in range(training_iter):
            start = time.time()
            # Zero gradients from previous iteration
            optimizer.zero_grad()
            # Output from model
            output = self(train_x)
            # Calculate loss and backpropagate gradients
            loss = -mll(output, train_y)
            loss.backward()
            end = time.time()
            if verbose:
                print('Iter %d/%d - Loss: %.3f' % (
                    i + 1, training_iter, loss.item()), 'in ', str(end - start))
            optimizer.step()

    def predict(self, x, likelihood, full_cov=False):
        self.eval()
        likelihood.eval()
        with torch.no_grad(), gpytorch.settings.fast_pred_var():
            # https://github.com/cornellius-gp/gpytorch/issues/728
            observed_pred = likelihood(self(x))
            if not full_cov:
                mean = observed_pred.mean
                var = observed_pred.variance
            else:
                mean = observed_pred.mean
                var = observed_pred.covariance_matrix
        return mean, var


# Sparse GP
train_x = torch.stack((
    torch.linspace(0, 1, 100), torch.linspace(0, 1, 100)), dim=1)
# True function is sin(2*pi*x) with Gaussian noise
train_y = torch.sin(train_x[:, 0] * (2 * math.pi)) + torch.randn(
    train_x[:, 0].size()) * math.sqrt(0.04)
# initialize likelihood, kernel and model
likelihood = gpytorch.likelihoods.GaussianLikelihood(
    noise_prior=gpytorch.priors.NormalPrior(0.04, 1))
likelihood.noise = torch.tensor(0.04)  # init noise var
lengthscale_prior = gpytorch.priors.MultivariateNormalPrior(
    torch.tensor([0.2, 0.2]),
    torch.diag(torch.tensor([6.0, 6.0])))
outputscale_prior = gpytorch.priors.NormalPrior(5.0, 10.)
hypers = {'base_kernel.lengthscale': lengthscale_prior.mean,
          'outputscale': outputscale_prior.mean}
kernel = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(
    ard_num_dims=train_x.shape[1], lengthscale_prior=lengthscale_prior),
    outputscale_prior=outputscale_prior)
kernel.initialize(**hypers)  # init kernel hyperparams
random_idx = torch.randint(0, len(train_x), size=(30,))
inducing_points = train_x[random_idx]
model = SparseGPModel(train_x, train_y, likelihood, kernel,
                      inducing_points=inducing_points)
# Evaluate GP by making predictions on test set
test_x = torch.stack((
    torch.linspace(0, 1, 51), torch.linspace(0, 1, 51)), dim=1)
start = time.time()
mean, var = model.predict(test_x, likelihood)
end = time.time()
print('gpytorch prediction time', end - start)
lower = mean - 2 * torch.sqrt(var)
upper = mean + 2 * torch.sqrt(var)
mean = mean
# Initialize plot
f, ax1 = plt.subplots(1, 1)
ax1.plot(train_x.numpy()[:, 0], train_y.numpy(), 'k*')
ax1.plot(test_x.numpy()[:, 0], mean.numpy(), 'b')
ax1.fill_between(test_x.numpy()[:, 0], lower.numpy(), upper.numpy(),
                 facecolor='b', alpha=0.5)
ax1.set_ylim([-3, 3])
ax1.set_title('GPyTorch')
ax1.legend(['Observed Data', 'Mean', 'Confidence'])
plt.show()
# Find optimal model hyperparameters
model.optimize(likelihood, train_x, train_y, optimizer=torch.optim.SGD,
               lr=0.01, training_iter=100)
# Evaluate GP by making predictions on test set
start = time.time()
mean, var = model.predict(test_x, likelihood)
end = time.time()
print('gpytorch prediction time', end - start)
lower = mean - 2 * torch.sqrt(var)
upper = mean + 2 * torch.sqrt(var)
# Initialize plot
f, ax1 = plt.subplots(1, 1)
ax1.plot(train_x.numpy()[:, 0], train_y.numpy(), 'k*')
ax1.plot(test_x.numpy()[:, 0], mean.numpy(), 'b')
ax1.fill_between(test_x.numpy()[:, 0], lower.numpy(), upper.numpy(),
                 facecolor='b', alpha=0.5)
ax1.set_ylim([-3, 3])
ax1.set_title('GPyTorch')
ax1.legend(['Observed Data', 'Mean', 'Confidence'])
plt.show()
```

**Comment by monabf** [comment]
Hi @gpleiss, any news on this? I can't seem to understand why the predictive variance is so small on this simple example and why it starts out with a numerical warning.
Thanks a lot :)

**Comment by gpleiss** [comment]
@monabf taking a look today

**Comment by gpleiss** [comment]
@monabf sorry for the slow work on this. I tried your code example, and I cannot reproduce your error. Here's what I'm seeing (same behavior for 10 runs) - variance is perfectly fitting the data.

Screen Shot 2021-04-05 at 7 59 09 AM

**Comment by monabf** [comment]
No worries @gpleiss, thanks for taking a look. That is so weird. After doing `pip uninstall gpytorch` and `pip install git+https://github.com/cornellius-gp/gpytorch.git@faster_sgpr_covar`, when I run said script and run model.optimize(), this is what I obtain every single time
Bildschirmfoto 2021-04-06 um 09 38 34

with a message about rounding up variances to 1e-10. Did you close the first plot to optimize the inducing points?

**Comment by gpleiss** [comment]
Ah sorry @monabf, I was on a different PR branch when I made that plot (my bad!)
I see the issue, and I'm looking into it now.

**Comment by gpleiss** [comment]
@monabf try now!

**Comment by monabf** [comment]
No worries @gpleiss, it looks good now!

File: gpytorch/models/exact_prediction_strategies.py [hunk 1]
```diff
@@ -637,7 +638,18 @@ def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_
     def exact_prediction(self, joint_mean, joint_covar):
         # Find the components of the distribution that contain test data
         test_mean = joint_mean[..., self.num_train :]
-        test_test_covar = joint_covar[..., self.num_train :, self.num_train :].evaluate_kernel()
+
+        # If we're in lazy evaluation mode, let's use the base kernel of the SGPR output to compute the prior covar
+        test_test_covar = joint_covar[..., self.num_train :, self.num_train :]
+        if isinstance(test_test_covar, LazyEvaluatedKernelTensor) and hasattr(test_test_covar.kernel, "base_kernel"):
```
Comment (jacobrgardner) [inline gpytorch/models/exact_prediction_strategies.py:644]: In 6 months is this going to be the next `if "variational" in param_name:`?

**Review by jacobrgardner — APPROVED** [review]
Looks good to me other than the string comparison

**Comment by gpleiss** [comment]
@jacobrgardner string comparison fixed. I'll merge once a final set of tests passes.

**Pull request merged**